pkg/uprobetracer/tracer.go

```diff
@@ -0,0 +1,326 @@
+)
+
+type inodeUUID struct {
+	token uint64
+}
+
+func getInodeUUID(file *os.File) (inodeUUID, error) {
+	// TODO: use `kfilefields` to get a unique token for each inode,
+	// here we just use `fd` as token, making each file has a distinct token.
+	return inodeUUID{uint64(file.Fd())}, nil
+}
+
+// inodeKeeper holds a file object, with the counter representing its
+// reference count. The link is not nil only when the file is attached.
+type inodeKeeper struct {
+	counter int
+	file    *os.File
+	link    link.Link
+}
+
+func (t *inodeKeeper) close() {
+	if t.link != nil {
+		t.link.Close()
+	}
+	t.file.Close()
+}
+
+type Tracer[Event any] struct {
+	progName       string
+	progType       ProgType
+	attachFilePath string
+	attachSymbol   string
+	prog           *ebpf.Program
+
+	// keeps the inodes for each attached container
+	// when users write library names in ebpf section names, it's possible to
+	// find multiple libraries of different archs within the same container,
+	// making this a one-to-many mapping
+	containerPid2Inodes map[uint32][]inodeUUID
+	// keeps the fd and refCount for each inodeUUID
+	inodeRefCount map[inodeUUID]*inodeKeeper
+	// used as a set, keeps PIDs of the pending containers
+	pendingContainerPids map[uint32]bool
+
+	logger logger.Logger
+
+	closed bool
+	mu     sync.Mutex
+}
+
+func NewTracer[Event any](logger logger.Logger) (*Tracer[Event], error) {
+	t := &Tracer[Event]{
+		containerPid2Inodes:  make(map[uint32][]inodeUUID),
+		inodeRefCount:        make(map[inodeUUID]*inodeKeeper),
+		pendingContainerPids: make(map[uint32]bool),
+		logger:               logger,
+		closed:               false,
+	}
+	return t, nil
+}
+
+// AttachProg loads the ebpf program, and try attaching if there are pending containers
+func (t *Tracer[Event]) AttachProg(progName string, progType ProgType, attachTo string, prog *ebpf.Program) error {
+	if progType != ProgUprobe && progType != ProgUretprobe {
+		return fmt.Errorf("unsupported uprobe prog type: %q", progType)
+	}
+
+	if prog == nil {
+		return errors.New("prog does not exist")
+	}
+	if t.prog != nil {
+		return errors.New("loading uprobe program twice")
+	}
+
+	parts := strings.Split(attachTo, ":")
+	if len(parts) < 2 {
+		return fmt.Errorf("invalid section name %q", attachTo)
+	}
+	if !filepath.IsAbs(parts[0]) && strings.Contains(parts[0], "/") {
+		return fmt.Errorf("section name must be either an absolute path or a library name: %q", parts[0])
+	}
+
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if t.closed {
+		return errors.New("uprobetracer has been closed")
+	}
+
+	t.progName = progName
+	t.progType = progType
+	t.attachFilePath = parts[0]
+	t.attachSymbol = parts[1]
+	t.prog = prog
+
+	// attach to pending containers, then release the pending list
+	for pid := range t.pendingContainerPids {
+		t.attach(pid)
+	}
+	t.pendingContainerPids = nil
+
+	return nil
+}
+
+func searchForLibrary(containerPid uint32, filePath string) ([]string, error) {
+	var libraryPaths []string
```

Comment:
Since it contains both executables and libraries, I suggest to rename this variable:
```suggestion
	var targetPaths []string
```

Ditto for the `libraryPath` variable in the for loop, it could be targetPath.